File: clients/Go/model/player_view.go
package model

import "io"
import . "aicup2020/stream"

type PlayerView struct {
    MyId int32
    MapSize int32
    FogOfWar bool
    EntityProperties map[EntityType]EntityProperties
    MaxTickCount int32
    MaxPathfindNodes int32
    CurrentTick int32
    Players []Player
    Entities []Entity
}
func NewPlayerView(myId int32, mapSize int32, fogOfWar bool, entityProperties map[EntityType]EntityProperties, maxTickCount int32, maxPathfindNodes int32, currentTick int32, players []Player, entities []Entity) PlayerView {
    return PlayerView {
        MyId: myId,
        MapSize: mapSize,
        FogOfWar: fogOfWar,
        EntityProperties: entityProperties,
        MaxTickCount: maxTickCount,
        MaxPathfindNodes: maxPathfindNodes,
        CurrentTick: currentTick,
        Players: players,
        Entities: entities,
    }
}
func ReadPlayerView(reader io.Reader) PlayerView {
    result := PlayerView {}
    result.MyId = ReadInt32(reader)
    result.MapSize = ReadInt32(reader)
    result.FogOfWar = ReadBool(reader)
    EntityPropertiesSize := ReadInt32(reader)
    result.EntityProperties = make(map[EntityType]EntityProperties)
    for i := int32(0); i < EntityPropertiesSize; i++ {
        var EntityPropertiesKey EntityType
        EntityPropertiesKey = ReadEntityType(reader)
        var EntityPropertiesValue EntityProperties
        EntityPropertiesValue = ReadEntityProperties(reader)
        result.EntityProperties[EntityPropertiesKey] = EntityPropertiesValue
    }
    result.MaxTickCount = ReadInt32(reader)
    result.MaxPathfindNodes = ReadInt32(reader)
    result.CurrentTick = ReadInt32(reader)
    result.Players = make([]Player, ReadInt32(reader))
    for i := range result.Players {
        result.Players[i] = ReadPlayer(reader)
    }
    result.Entities = make([]Entity, ReadInt32(reader))
    for i := range result.Entities {
        result.Entities[i] = ReadEntity(reader)
    }
    return result
}
func (value PlayerView) Write(writer io.Writer) {
    WriteInt32(writer, value.MyId)
    WriteInt32(writer, value.MapSize)
    WriteBool(writer, value.FogOfWar)
    WriteInt32(writer, int32(len(value.EntityProperties)))
    for EntityPropertiesKey, EntityPropertiesValue := range value.EntityProperties {
        WriteInt32(writer, int32(EntityPropertiesKey))
        EntityPropertiesValue.Write(writer)
    }
    WriteInt32(writer, value.MaxTickCount)
    WriteInt32(writer, value.MaxPathfindNodes)
    WriteInt32(writer, value.CurrentTick)
    WriteInt32(writer, int32(len(value.Players)))
    for _, PlayersElement := range value.Players {
        PlayersElement.Write(writer)
    }
    WriteInt32(writer, int32(len(value.Entities)))
    for _, EntitiesElement := range value.Entities {
        EntitiesElement.Write(writer)
    }
}

File: clients/Python-PyPy/model/player_view.py
from .entity_type import EntityType
from .entity_properties import EntityProperties
from .player import Player
from .entity import Entity
class PlayerView:
    def __init__(self, my_id, map_size, fog_of_war, entity_properties, max_tick_count, max_pathfind_nodes, current_tick, players, entities):
        self.my_id = my_id
        self.map_size = map_size
        self.fog_of_war = fog_of_war
        self.entity_properties = entity_properties
        self.max_tick_count = max_tick_count
        self.max_pathfind_nodes = max_pathfind_nodes
        self.current_tick = current_tick
        self.players = players
        self.entities = entities
    @staticmethod
    def read_from(stream):
        my_id = stream.read_int()
        map_size = stream.read_int()
        fog_of_war = stream.read_bool()
        entity_properties = {}
        for _ in range(stream.read_int()):
            entity_properties_key = EntityType(stream.read_int())
            entity_properties_value = EntityProperties.read_from(stream)
            entity_properties[entity_properties_key] = entity_properties_value
        max_tick_count = stream.read_int()
        max_pathfind_nodes = stream.read_int()
        current_tick = stream.read_int()
        players = []
        for _ in range(stream.read_int()):
            players_element = Player.read_from(stream)
            players.append(players_element)
        entities = []
        for _ in range(stream.read_int()):
            entities_element = Entity.read_from(stream)
            entities.append(entities_element)
        return PlayerView(my_id, map_size, fog_of_war, entity_properties, max_tick_count, max_pathfind_nodes, current_tick, players, entities)
    def write_to(self, stream):
        stream.write_int(self.my_id)
        stream.write_int(self.map_size)
        stream.write_bool(self.fog_of_war)
        stream.write_int(len(self.entity_properties))
        for key, value in self.entity_properties.items():
            stream.write_int(key)
            value.write_to(stream)
        stream.write_int(self.max_tick_count)
        stream.write_int(self.max_pathfind_nodes)
        stream.write_int(self.current_tick)
        stream.write_int(len(self.players))
        for element in self.players:
            element.write_to(stream)
        stream.write_int(len(self.entities))
        for element in self.entities:
            element.write_to(stream)
    def __repr__(self):
        return "PlayerView(" + \
            repr(self.my_id) + "," + \
            repr(self.map_size) + "," + \
            repr(self.fog_of_war) + "," + \
            repr(self.entity_properties) + "," + \
            repr(self.max_tick_count) + "," + \
            repr(self.max_pathfind_nodes) + "," + \
            repr(self.current_tick) + "," + \
            repr(self.players) + "," + \
            repr(self.entities) + \
            ")"
